sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala

```diff
@@ -61,4 +65,202 @@ class CatalystTypeConvertersSuite extends SparkFunSuite {
+    val d = D(B(A(1), 2.0), Seq(A(1), A(5)), Seq(3, 8), 10L)
+    assert(d === rebuildWithProductConverter(d))
+  }
+
+  test("createToProductConverter[T] with T having Maps") {
+    val e = E(B(A(1), 2.0),
+      Map(A(1) -> A(5), A(11) -> A(15)),
+      Map(A(1) -> 5, A(11) -> 15),
+      Map(1 -> A(5), 11 -> A(15)),
+      Map(1 -> 5, 11 -> 15))
+    assert(e === rebuildWithProductConverter(e))
+  }
+
+  test("createToProductConverter[T] with T having multiple constructors") {
+    val f = new F(1)
+    assert(F(1, "hi everybody") === rebuildWithProductConverter(f))
+  }
+
+  test("createToProductConverter[T] with T having String") {
+    val g = new G("hi everybody!")
+    assert(g === rebuildWithProductConverter(g))
+  }
+
+  test("createToProductConverter[T] with an incompatible case class fails at conversion") {
+    val a = A(1)
+    val dataType = schemaFor[A].dataType
+    val row = CatalystTypeConverters.createToCatalystConverter(dataType)(a)
+    val converter =
+      CatalystTypeConverters.createToProductConverter[G](dataType.asInstanceOf[StructType])
+    intercept[ClassCastException] { converter(row.asInstanceOf[InternalRow]) }
+  }
+
+  test("createToProductConverter[T] with T having Some[U]") {
+    val h = H(Some(1))
+    assert(h === rebuildWithProductConverter(h))
+  }
+
+  test("createToProductConverter[T] with T having None") {
+    val h = H(None)
+    assert(h === rebuildWithProductConverter(h))
+  }
+
+  test("createToProductConverter[T] with T having Option[Option[U]]") {
+    val i = I(Some(H(Some(1))))
+    assert(i === rebuildWithProductConverter(i))
+  }
+
+  test("createToProductConverter[T] with T having Option[Seq[U]]") {
+    val j = J(Some(Seq(A(1), A(2))))
+    assert(j === rebuildWithProductConverter(j))
+  }
+
+  test("createToProductConverter[T] with T having BigDecimal") {
+    val k = K(BigDecimal("123.0"))
+    assert(k === rebuildWithProductConverter(k))
+  }
+
+  test("createToProductConverter[T] with T having java.math.BigDecimal") {
+    // NOTE: As currently implemented, Decimal.apply(java.math.BigDecimal) triggers the implicit
+    // BigDecimal.javaBigDecimal2bigDecimal, creating a Scala BigDecimal with
+    // BigDecimal.defaultMathContext. So a given java.math.BigDecimal ends up taking maximum
+    // precision, and the scale is truncated to Decimal.MAX_LONG_DIGITS when converting back. This
+    // unit test accounts for that current behavior.
+    val l = L(new java.math.BigDecimal(new java.math.BigInteger("123"), Decimal.MAX_LONG_DIGITS))
+    assert(l === rebuildWithProductConverter(l))
+  }
+
+  test("createToProductConverter[T] with T having java.sql.Date") {
+    val m = M(DateTimeUtils.toJavaDate(daysSinceEpoch = 16000))
+    assert(m === rebuildWithProductConverter(m))
+  }
+
+  test("createToProductConverter[T] with T having java.sql.Timestamp") {
+    val n = N(DateTimeUtils.toJavaTimestamp(System.currentTimeMillis * 1000L))
+    assert(n === rebuildWithProductConverter(n))
+  }
```

Comment:
I didn't have a test for timestamps in the previous versions of this PR, so this is new.